common/short-hex-str/src/lib.rs

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use mirai_annotations::debug_checked_precondition;
+use serde::{Serialize, Serializer};
+use std::{fmt, str};
+use thiserror::Error;
+
+/// An efficient container for formatting a byte slice as a hex-formatted string,
+/// stored on the stack.
+///
+/// Using `ShortHexStr` instead of `hex::encode` is about 3-4x faster on a recent
+/// MBP 2019 (~48 ns/iter vs ~170 ns/iter) in an artifical micro benchmark.
+#[derive(Ord, PartialOrd, Eq, PartialEq, Hash, Clone, Copy)]
+pub struct ShortHexStr([u8; ShortHexStr::LENGTH]);
+
+#[derive(Error, Debug)]
+#[error("Input bytes are too short")]
+pub struct InputTooShortError;
+
+impl ShortHexStr {
+    pub const SOURCE_LENGTH: usize = 4;
+    pub const LENGTH: usize = 2 * ShortHexStr::SOURCE_LENGTH;
+
+    /// Format a new `ShortHexStr` from a byte slice.
+    ///
+    /// Returns `Err(InputTooShortError)` if the input byte slice length is less
+    /// than `SOURCE_LENGTH` bytes.
+    pub fn from_bytes(src_bytes: &[u8]) -> Result<ShortHexStr, InputTooShortError> {
+        if src_bytes.len() >= ShortHexStr::SOURCE_LENGTH {
+            Ok(ShortHexStr::from_bytes_unchecked(src_bytes))
+        } else {
+            Err(InputTooShortError)
+        }
+    }
+
+    /// Format a new `ShortHexStr` from a byte slice.
+    ///
+    /// Panics if the input byte slice length is less than `SOURCE_LENGTH` bytes.
+    pub fn from_bytes_unchecked(src_bytes: &[u8]) -> Self {
+        debug_checked_precondition!(src_bytes.len() >= ShortHexStr::SOURCE_LENGTH);
+
+        let src_short_bytes = &src_bytes[0..ShortHexStr::SOURCE_LENGTH];
+        let mut dest_bytes = [0u8; ShortHexStr::LENGTH];
+
+        // We include a tiny hex encode here instead of using the `hex` crate's
+        // `encode_to_slice`, since the compiler seems unable to inline across
+        // the crate boundary.
+        hex_encode(&src_short_bytes, &mut dest_bytes);
+        Self(dest_bytes)
```

Comment:
why store this in bytes when we're always going to use this as a string? (unless I'm missing smthg)

Reply:
So `&str` is basically a `&[u8]` with an additional invariant that it must only contain valid UTF-8. Unfortunately, there's no way other way to represent an owned, stack-allocated string than a `[u8; N]` : (. The usual `String` type is always heap allocated; it's more-or-less a `Box<[u8]>` with some extra invariants.

So we have to store a `[u8; N]` and then cast its `&[u8]` to a `&str`. In `as_str()` you can see the `str::from_utf8(&self.0)` which just checks that the slice contains all valid UTF-8 and then casts the `&[u8]` to a `&str`. If we wanted to be extra (shave off 6-7 ns overhead lol) we could use `unsafe { str::from_utf8_unchecked(&self.0) }`, which literally just casts the refs with no check. Either way it's safe since the hex_encode will only ever emit valid UTF-8.

Reply:
ah I see, makes sense!